pkg/asset/installconfig/openstack/session.go

```diff
@@ -40,49 +40,63 @@ func defaultClientOpts(cloudName string) *clientconfig.ClientOpts {
 type yamlLoadOpts struct{}
 
 func (opts yamlLoadOpts) LoadCloudsYAML() (map[string]clientconfig.Cloud, error) {
-	return loadAndLog(clientconfig.FindAndReadCloudsYAML)
+	var clouds clientconfig.Clouds
+	content, err := loadAndLog(clientconfig.FindAndReadCloudsYAML)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(content, &clouds)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal yaml")
+	}
+
+	return clouds.Clouds, nil
 }
 
 func (opts yamlLoadOpts) LoadSecureCloudsYAML() (map[string]clientconfig.Cloud, error) {
-	clouds, err := loadAndLog(clientconfig.FindAndReadSecureCloudsYAML)
+	var clouds clientconfig.Clouds
+	content, err := loadAndLog(clientconfig.FindAndReadSecureCloudsYAML)
 	if err != nil {
 		if err.Error() == "no secure.yaml file found" {
 			// secure.yaml is optional so just ignore read error
-			return clouds, nil
+			return clouds.Clouds, nil
 		}
 	}
-	return clouds, err
+	err = yaml.Unmarshal(content, &clouds)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal yaml")
+	}
+	return clouds.Clouds, err
 }
 
 func (opts yamlLoadOpts) LoadPublicCloudsYAML() (map[string]clientconfig.Cloud, error) {
-	clouds, err := loadAndLog(clientconfig.FindAndReadPublicCloudsYAML)
+	var publicClouds clientconfig.PublicClouds
+	content, err := loadAndLog(clientconfig.FindAndReadPublicCloudsYAML)
 	if err != nil {
 		if err.Error() == "no clouds-public.yaml file found" {
 			// clouds-public.yaml is optional so just ignore read error
-			return clouds, nil
+			return publicClouds.Clouds, nil
 		}
 	}
-	return clouds, err
+	err = yaml.Unmarshal(content, &publicClouds)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal yaml")
+	}
+	return publicClouds.Clouds, err
 }
 
-func loadAndLog(fn func() (string, []byte, error)) (map[string]clientconfig.Cloud, error) {
+func loadAndLog(fn func() (string, []byte, error)) ([]byte, error) {
```

Comment:
Why are we changing the return type to []byte? All LoadXCloudsYaml() functions return map[string]Cloud

Reply:
Now in `loadAndLog` function we always unmarshal data to `clientconfig.Clouds` type, but this is wrong, because for `clouds-public.yaml` we must use `clientconfig.PublicClouds`, otherwise the result will be empty and the installation fail.
To avoid this I decided to keep only loading of content and logging in `loadAndLog`, and move the unmarshaling to the upper level.
In `LoadCloudsYAML` and `LoadSecureCloudsYAML` we continue to unmarshal to `clientconfig.Clouds`, but for `LoadPublicCloudsYAML` we now use `clientconfig.PublicClouds`

Reply:
ah ok, I follow now